userspace/libsinsp/parsers.cpp

```diff
@@ -4522,38 +4536,61 @@ void sinsp_parser::parse_setpgid_exit(sinsp_evt *evt)
 		pgid = *(int64_t *)parinfo->m_val;
 		ASSERT(parinfo->m_len == sizeof(int64_t));
 
-		//
-		// If pid is zero, then the process ID of the calling process is used.
-		//
-		sinsp_threadinfo* ptinfo = NULL;
-		if (pid == 0)
+		// Determine whether the caller is in a pid namespace
+		// or not. That determines how we should interpret the
+		// pid arg.
+		if(cinfo->m_pid == cinfo->m_vpid)
 		{
-			ptinfo = evt->get_thread_info();
+			// The caller is not in a pid namespace, so
+			// the pid is already in the global pid
+			// namespace.
+			gpid = pid;
 		}
 		else
 		{
-			ptinfo = m_inspector->get_thread(pid, false, true);
+			// pidarg is relative to the namespace of the
+			// caller. Convert it to the global namespace.
+			gpid = (pid == 0 ?
+				0 :
+				m_inspector->m_thread_manager->get_pid_for_vpid(cinfo->m_container_id, pid));
+
+			if(gpid == -1)
+			{
+				ASSERT(false);
```

Comment:
I still hit this assertion almost 100% of the times. The assertion seems to be assuming that the target of setpgid is already in the thread table by the time this gets called, however, with containers that's not the case:

```
85918 00:58:16.211933897 3 elated_mclean (52808fbc9bcd) bash (23370:1) > clone
85923 00:58:16.212132529 3 elated_mclean (52808fbc9bcd) bash (23370:1) < clone res=34 exe=/bin/bash args= tid=23370(bash) pid=23370(bash) ptid=23350(docker-containe) cwd= fdlimit=1048576 pgft_maj=31 pgft_min=1367 vm_size=18236 vm_rss=3152
 vm_swap=0 comm=bash cgroups=cpuset=/docker/52808fbc9bcd4c1cd342375b32f41f2b9e0f14760e808b15d936a79dc5cd0e... flags=25165824(CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID) uid=0 gid=0 vtid=1(init) vpid=1(init)
85925 00:58:16.212198210 3 elated_mclean (52808fbc9bcd) bash (23370:1) > setpgid pid=34 pgid=34

---> here you call get_pid_for_vpid, but thread 34 hasn't been created yet, because we don't know its global tid. So, the assertion triggers

85933 00:58:16.212246020 4 elated_mclean (52808fbc9bcd) bash (26297:34) < clone res=0 exe=/bin/bash args= tid=26297(bash) pid=26297(bash) ptid=23370(bash) cwd= fdlimit=1048576 pgft_maj=0 pgft_min=1 vm_size=18236 vm_rss=468 vm_swap=0 comm=
bash cgroups=cpuset=/docker/52808fbc9bcd4c1cd342375b32f41f2b9e0f14760e808b15d936a79dc5cd0e... flags=25165824(CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID) uid=0 gid=0 vtid=34 vpid=34

---> here thread "34" gets created, since we know for the first time its global tid (26297)
```

Reply:
So you had a test program that did a `clone()` and then a `setpgid()` immediately after, right?

Reply:
Not a test program, my docker does it 100% of the time with:

```
$ docker run -it --rm ubuntu
# ls
```